evm/spec/framework.tex

```diff
@@ -33,5 +33,119 @@ \subsection{Field selection}
+This ensures that the columns in $S_1'$ and the columns in $S_2'$ are permutations of each other.
+
+To sum up, for each STARK $S$, the prover:
+\begin{enumerate}
+  \item constructs a running product $P_i^l$ for each table looking into $S$ (called looking products here),
+  \item constructs a running product $P^S$ for $S$ (called looked product here),
+  \item sends the final value for each running product $P_{i, j}^l$ and $P^S_1$ to the verifier,
+  \item sends a commitment to $P_i^l$  and $P^S$ to the verifier.
+\end{enumerate}
+Then, for each STARK $S$, the verifier:
+\begin{enumerate}
+  \item computes the product $P = \prod_i P_{i, l}^l$,
+  \item checks that $P = P^S_1$,
+  \item checks that each $P_i^l$  and $P^S$ was correctly constructed.
+\end{enumerate}
+
+\subsection{Range-checks}
+\label{rc}
+In most cases, tables deal with U256 words, split into 32-bit limbs (to avoid overflowing the field). To prevent a malicious prover from cheating, it is crucial to range-check those limbs. 
+\subsubsection{What to range-check?}
+One can note that every element that ever appears on the stack has been pushed. Therefore, enforcing a range-check on pushed elements is enough to range-check all elements on the stack. Similarly, all elements in memory must have been written prior, and therefore it is enough to range-check memory writes. However, range-checking the PUSH and MSTORE opcodes is not sufficient.
+\begin{enumerate}
+  \item Pushes and memory writes are range-checked in ``BytePackingStark''.
+  \item Syscalls, exceptions and prover inputs are range-checked in ``ArithmeticStark''.
+  \item The inputs and outputs of binary and ternary arithmetic operations are range-checked in ``ArithmeticStark''.
+  \item The inputs' bits of logic operations are checked to be either 1 or 0 in ``LogicStark''. Since ``LogicStark'' only deals with bitwise operations, this is enough to have range-checked outputs as well.
+  \item The inputs of Keccak operations are range-checked in ``KeccakStark''. The output digest is written as bytes in ``KeccakStark''. Those bytes are used to reconstruct the associated 32-bit limbs checked against the limbs in ``CpuStark''. This implictly ensures that the output is range-checked.
+\end{enumerate}
+Note that some operations do not require a range-check:
+\begin{enumerate}
+  \item ``EQ'' reads two -- already range-checked -- elements on the stack, and checks they are equal. The output is either 0 or 1, and does therefore not need to be checked.
+  \item ``NOT'' reads one -- already range-checked -- element. The result is constrained to be equal to $\texttt{0xFFFFFFFF} - \texttt{input}$, which implicitly enforces the range check.
+  \item ``PC'': the program counter cannot be greater than $2^{32}$ in user mode. Moreover, in kernel mode, one can only jump to valid jump destinations. These two points implicitly enforce $PC$'s range check.
```

Comment:
> cannot be greater than $2^{32}$ in user mode

Maybe explain why (that it's not feasible to have user code longer than 2^32, and that a jump to anything else than a JUMPDEST would fail because of constraints).

> Moreover, in kernel mode, one can only jump to valid jump destinations.

I find it unclear. I would rather say that every jump in the kernel is made to a location in the kernel, and that the kernel code is smaller than 2^32.